File: ReorderDynamicContent/ReorderDialogCustom.js
// <reference name="MicrosoftAjax.js"/>
Type.registerNamespace("SitefinityWebApp.ReorderDynamicContent");

SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom = function (element) {

    SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom.initializeBase(this, [element]);
};

SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom.prototype = {

    /* --------------------------------- set up and tear down --------------------------------- */

    initialize: function () {
        SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom.callBaseMethod(this, 'initialize');
    },

    dispose: function () {
        SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom.callBaseMethod(this, 'dispose');
    },

    prepareDialog: function (dataItem, params) {
        // no folder param
        params["folderId"] = "";
        // no parent call needed
        if (!dataItem.SystemParentType) {
            dataItem.Id = "";           
            parentId = "";
        } // set the parent id to take advantage of the default logic
        else {
            dataItem.ParentId = dataItem.SystemParentId;
        }

        // base logic
        SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom.callBaseMethod(this, 'prepareDialog', [dataItem, params]);
        // set the dialog title
        this._dialogTitle.innerHTML = String.format(this._dialogTitlePattern, params["itemNameWithArticle"] || "Items");
    },

    _saveChanges: function () {
        var result = [];
        var itemElements = this._imageBinder.get_currentItemElements();
        for (var i = 0, length = itemElements.length; i < length; i++) {
            var element = itemElements[i];
            var dataItem = jQuery(element).data("dataItem");
            var ordinal = dataItem.Ordinal;
            var initialOrdinal = jQuery(element).data("initialOrdinal");

            if (initialOrdinal != null && initialOrdinal != dataItem.Ordinal) {
                result.push({ Key: dataItem.Id, Value: dataItem.Ordinal });
            }
        }
        
        var clientManager = this._imageBinder.get_manager();
        var serviceUrl = this._webSaveServiceUrl + "/ReorderDynamicContent/";
        var urlParams = [];

        if (this._itemType != null) {
            urlParams["itemType"] = this._itemType;
        }

        if (this._providerName) {
            urlParams["provider"] = this._providerName;
        }

        var keys = [];
        var data = result;
        clientManager.InvokePut(serviceUrl, urlParams, keys, data, this._saveChangesSuccess, this._saveChangesFailure, this);
    }
};

SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom.registerClass('SitefinityWebApp.ReorderDynamicContent.ReorderDialogCustom',
    Telerik.Sitefinity.Modules.Libraries.Web.UI.ReorderDialog);

File: ReorderDynamicContent/ReorderDialogCustom.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.UI;
using Telerik.Sitefinity.DynamicModules.Model;
using Telerik.Sitefinity.Localization;
using Telerik.Sitefinity.Modules.Libraries;
using Telerik.Sitefinity.Modules.Libraries.Web.UI;
using Telerik.Sitefinity.Web.UI;

namespace SitefinityWebApp.ReorderDynamicContent
{
    public class ReorderDialogCustom : ReorderDialog
    {
        protected override void InitializeControls(Telerik.Sitefinity.Web.UI.GenericContainer container)
        {
            // ItemType should be resolved from the dialog parameters for each module
            if (this.ItemType.BaseType != null && this.ItemType.BaseType.FullName == typeof(DynamicContent).FullName)
            {
                var binderContainer = this.ImageBinder.Containers.Where<BinderContainer>(c => c.ID == "BinderContainer1").FirstOrDefault();
                if (binderContainer != null)
                {
                    binderContainer.Markup =
                        @"<span sys:title=""{{Title.PersistedValue}}"" style=""border:1px solid #000"" >{{(Title.PersistedValue.length > 22) ? Title.PersistedValue.substring(0,22) + '...' : Title.PersistedValue}}</span>";
                }

                this.ImageBinder.HandleItemReordering = true;
                this.ImageBinder.AutoUpdateOrdinals = false;
                this.ImageBinder.ServiceUrl = this.GetServiceUrl();
                this.ImageBinder.ServiceChildItemsBaseUrl = this.GetChildServiceUrl();
                this.BackButton.Text = string.Format(Res.Get<Labels>().BackToAllItemsParameter, this.ItemsName);
                DragToChangeOrderLiteral.Text = string.Format(Res.Get<LibrariesResources>().DragToChangeOrder, this.ItemNameWithArticle);
            }
            else
            {
                base.InitializeControls(container);
            }
        }

        protected virtual string GetChildServiceUrl()
        {
            return ReorderDialogCustom.childServiceUrl;
        }

        protected virtual string GetServiceUrl()
        {
            return ReorderDialogCustom.serviceUrl;
        }

        public override IEnumerable<ScriptReference> GetScriptReferences()
        {
            var scripts = base.GetScriptReferences().ToList();
            scripts.Add(new ScriptReference(ReorderDialogCustom.dialogScript));
            return scripts;
        }

        public override IEnumerable<System.Web.UI.ScriptDescriptor> GetScriptDescriptors()
        {
            var @base = base.GetScriptDescriptors();

            var descriptor = (ScriptControlDescriptor)@base.Last();
            descriptor.Type = typeof(ReorderDialogCustom).FullName;
            descriptor.AddProperty("_webServiceUrl", this.GetServiceUrl());
            descriptor.AddProperty("_webSaveServiceUrl", ReorderDialogCustom.reorderDynamicContentServiceUrl);

            return @base;
        }

        internal const string dialogScript = "~/ReorderDynamicContent/ReorderDialogCustom.js";
        protected static readonly string reorderDynamicContentServiceUrl = "/Sitefinity/Services/ReorderDynamicContentService.svc";
        protected static readonly string serviceUrl = "/Sitefinity/Services/DynamicModules/Data.svc/?managerType=Telerik.Sitefinity.DynamicModules.DynamicModuleManager";
        protected static readonly string childServiceUrl = "/Sitefinity/Services/DynamicModules/Data.svc/parent/?managerType=Telerik.Sitefinity.DynamicModules.DynamicModuleManager";
    }
}